### DISCUSSION TITLE
isValid not updating correctly in my fancy wizard.

### DISCUSSION THREAD

**Discussion by fini** [react-hook-form/react-hook-form#9423] [Q&A] [answered]
I have a quite large and complex form that is using a wizard system, and I am dynamically assigning yupSchema for validation on each step.

In order for me to know if a user is allowed to move forward in the wizard, I call await trigger() in order for isValid to be updated. I then quickly clear the errors again, so the wizard step looks nice and clean when the user enters.

However for some reason my trigger does generate errors, but still doesn't update the isValid accordingly. I have found a workaround using a setTimeout to call doTrigger, but it does cause some visible flickering of the error messages being shown and removed.

I can also note that I have a dev-toolbar where I can do a trigger(), and hitting that immediately updates the isValid correctly, so yeah there is seemingly some timing thing going on here.

Any help would be greatly appreciated. ;)

```
  // Validate form when wizard step changes, and clear errors so the form looks clean 
  useEffect(() => {
    const doTrigger = async () => {
      await trigger()
      clearErrors()
    }

    setTimeout(() => doTrigger(), 0)
    // doTrigger()
  }, [activeStep, isValid, clearErrors, trigger])
```

**Comment by bluebill1049** [comment]
could you show a codesandbox?

**Reply by fini** [reply]
Sadly I don't see that as a possibility, it would take me days to do that, since the whole webapp is based on this form. :(

**Reply by bluebill1049** [reply]
which version are you running?

**Reply by fini** [reply]
7.39.5

**Reply by fini** [reply]
I tried running with 7.40.0-next.1, the issue was still present, but my error component didn't show anything.

it uses useFormContext to get the errors:

```
  const {
    formState: { errors },
  } = useFormContext()
```

**Reply by fini** [reply]
I also tried with 7.38.0, in which the issue is also present. I will try your debug snippet now 🙏🏻

**Comment by fini** [comment]
Here is a screenshot:
As can be seen in top left corner isValid is YES (true), but there are errors which I show in the yellow box, and 'Next Step' is not disabled like it should because it relies on isValid.
https://capture.dropbox.com/lZo9T0J2wobVvkqY

**Reply by bluebill1049** [reply]
can you run this code snippet: 

```tsx
resolver: async (data, context, options) => {
  // you can debug your validation schema here
  console.log('formData', data)
  console.log('validation result', await anyResolver(schema)(data, context, options))
  return anyResolver(schema)(data, context, options)
},
```

https://react-hook-form.com/api/useform

![Screenshot 2022-11-24 at 8 34 17 pm](https://user-images.githubusercontent.com/10513364/203748774-dacf0497-ed1c-4fb7-a080-e82072f6d139.png)

so you can debug what's the issue, 

- `isValid` will run the entire form
- `errors` is related to field level depend on your selection of `mode`

**Reply by fini** [reply]
Well this seems awesome, when using your snippet my bug disappears ;)

my resolver:
`resolver: yupResolver(schema)`

your snippet with yupResolver:
```
 resolver: async (data, context, options) => {
      // you can debug your validation schema here
      console.log('formData', data)
      console.log(
        'validation result',
        await yupResolver(schema)(data, context, options)
      )
      return yupResolver(schema)(data, context, options)
    }
```

**Reply by fini** [reply]
My colleague just reminded me that last week I implemented an async validation in YUP where we check emails on our backend API. So that could indeed be why this issue has arrived since don't do any async/await on the yupResolver.  🤗

**Reply by fini** [reply]
I can confirm that the issue was solely because of my AJAX call in my Yup Test:

```
      .test('is-email-valid', '${path} does not exist', async (value) =>
        testEmail(value)
      ),
```

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/discussions/9423#discussioncomment-4223951

**Reply by fini** [reply]
It seems it only works if I call yupResolver twice..?

```
    resolver: async (data, context, options) => {
      await yupResolver(schema)(data, context, options)
      const yupResult = await yupResolver(schema)(data, context, options)
      return yupResult
    },
```

Any suggestion?

Thanks for your help so far, I appreciate it very much!

**Reply by bluebill1049** [reply]
not sure, hook form only resolve what's returned by your schema result. can you try to log each schema return result and see the diff.

**Reply by bluebill1049** [reply]
Did you figure out the root cause?

**Reply by fini** [reply]
I have located the main issue, as mentioned it was related to a Yup async test:

**not working:**

```
.test('is-email-valid', '${path} does not exist', async (value) =>
  testEmail(value)
)
```
**working:**

```
.test('is-email-valid', '${path} does not exist', (value) =>
  testEmail(value)
)
```

I am now back to using resolver as usual:
`resolver: yupResolver(schema)`

**Reply by fini** [reply]
the testEmail function is using async/await

**Comment by bluebill1049** [comment, marked as answer]
https://github.com/react-hook-form/react-hook-form/discussions/9423#discussioncomment-4236065

https://github.com/react-hook-form/react-hook-form/discussions/9423#discussioncomment-4236075
